index.coffee

```diff
@@ -204,83 +233,174 @@ exports.DockerProgress = class DockerProgress
+			accumulatedSize = 0
+			_.forEachRight layers, (layer) ->
+				if layer.Id is '<missing>'
+					accumulatedSize += layer.Size
+				else
+					layer.Size += accumulatedSize
+					accumulatedSize = 0
+					concreteLayers.push(layer)
+			return concreteLayers
+		.map (layer) ->
+			# transform into something like this:
+			# [ {id: "bb12b...", size: 1339, blobs: [ "a694d3...", ... ]}, ... ]
+			d.getImage(layer['Id']).inspectAsync().then (info) ->
+				id: stripShaPrefix(layer['Id'])
+				size: layer['Size']
+				blobs: _(info['RootFS']['Layers'])
+					.map(stripShaPrefix)
+					.uniq()
+					.value()
+		.then (layers) ->
+			# each layer references all blobs it's made of. for each layer
+			# keep only layer-specific blobs, dropping older ones.
+			knownBlobs = null
+			_.forEach layers, (layer) ->
+				if not knownBlobs?
+					knownBlobs = layer.blobs
+					return
+				layer.blobs = _.dropWhile layer.blobs, (blob) ->
+					_.includes(knownBlobs, blob)
+				knownBlobs = knownBlobs.concat(layer.blobs)
+			return layers
+		.then (layers) ->
+			# distribute the layer size uniformly across each layer's blobs
+			# and return a map of all known blob IDs to their size.
+			_(layers)
+				.flatMap (layer) ->
+					blobCount = layer.blobs.length
+					return [] if blobCount is 0
+					blobSize = layer.size // blobCount
+					remainder = layer.size - blobSize * blobCount
+					_.map layer.blobs, (blobId, index) ->
+						[ blobId, blobSize + (if index is 0 then remainder else 0) ]
+				.fromPairs()
+				.value()
+
+	# Create a stream that transforms `docker.modem.followProgress` onProgress
+	# events to include total progress metrics.
 	pushProgress: (image, onProgress) ->
 		@getImageLayerSizes(image)
 		.then (layerSizes) ->
-			layerIds = _.keys(layerSizes)
-			layerPushedSize = {}
-			completedSize = 0
+			layerIds = {} # map from remote to local ids
+			remoteLayerIds = _.keys(layerSizes)
 			totalSize = _.sum(_.values(layerSizes))
+			completedSize = 0
+			lastPushedSize = 0
+			layerPushedSize = {}
 			return (evt) ->
 				try
 					{ status } = evt
 					shortId = evt.id
+					if evt.progressDetail? and not layerIds[shortId]?
+						remapShortId(shortId, remoteLayerIds, layerIds)
+
 					pushMatch = /Image (.*) already pushed/.exec(status)
+
 					if status is 'Pushing' and evt.progressDetail.current?
-						longId = getLongId(shortId, layerIds)
-						if longId?
-							layerPushedSize[longId] = evt.progressDetail.current
-					else if status is 'Layer already exists' or status is 'Image successfully pushed' or pushMatch
+						layerPushedSize[shortId] = evt.progressDetail.current
+					# registry2-only statuses
+					else if status is 'Pushed' or /^Mounted from /.test(status)
+						remoteId = layerIds[shortId]
+						completedSize += layerSizes[remoteId]
+						layerPushedSize[shortId] = 0
+					# registry1-only statuses
+					else if _.includes(['Already exists', 'Layer already exists', 'Image successfully pushed'], status) or pushMatch
```

Comment:
It'd be nicer to put the `pushMatch` first, that way it's easier to spot it (rather than a little dangling bit after a large first bit) and also makes use of the work already done there if it's a hit